frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -196,248 +219,351 @@ function ReaderBookmark:addToMainMenu(menu_items)
-                        if compare_xp then
-                            return compare_xp < 0
-                        end
-                        logger.warn("Invalid xpointer in highlight:", a.pos1, b.pos1)
-                        return
-                    end
-                    return a.highlighted -- have page bookmarks before highlights
-                end
-                return compare_xp < 0
-            end
-            -- if compare_xp is nil, some xpointer is invalid and will be sorted first to page 1
-            logger.warn("Invalid xpointer in highlight:", a.page, b.page)
+        pn_or_xp = self:getCurrentPageNumber()
+    end
+    local index = self:getDogearBookmarkIndex(pn_or_xp)
+    if index then
+        item = table.remove(self.ui.annotation.annotations, index)
+    else
+        local text
+        local chapter = self.ui.toc:getTocTitleByPage(pn_or_xp)
+        if chapter == "" then
+            chapter = nil
+        else
+            -- @translators In which chapter title (%1) a note is found.
+            text = T(_("in %1"), chapter)
         end
-        return a_page > b_page
+        item = {
+            page = pn_or_xp,
+            text = text,
+            chapter = chapter,
+        }
+        self.ui.annotation:addItem(item)
     end
+    self.ui:handleEvent(Event:new("AnnotationsModified", { item }))
+end
+
+function ReaderBookmark:setDogearVisibility(pn_or_xp)
+    local visible = self:isPageBookmarked(pn_or_xp)
+    self.view.dogear:onSetDogearVisibility(visible)
+end
+
+function ReaderBookmark:isPageBookmarked(pn_or_xp)
+    local page = pn_or_xp or self:getCurrentPageNumber()
+    return self:getDogearBookmarkIndex(page) and true or false
 end
 
 function ReaderBookmark:isBookmarkInPageOrder(a, b)
     local a_page = self:getBookmarkPageNumber(a)
     local b_page = self:getBookmarkPageNumber(b)
-    if a_page == b_page then -- have bookmarks before highlights
-        return a.highlighted
+    if a_page == b_page then -- have page bookmarks before highlights
+        return not a.drawer
+    end
+    return a_page < b_page
+end
+
+function ReaderBookmark:getDogearBookmarkIndex(pn_or_xp)
+    local doesMatch
+    if self.ui.paging then
+        doesMatch = function(p1, p2)
+            return p1 == p2
+        end
+    else
+        doesMatch = function(p1, p2)
+            return self.ui.document:getPageFromXPointer(p1) == self.ui.document:getPageFromXPointer(p2)
+        end
+    end
+    local _middle
+    local _start, _end = 1, #self.ui.annotation.annotations
+    while _start <= _end do
+        _middle = bit.rshift(_start + _end, 1)
+        local v = self.ui.annotation.annotations[_middle]
+        if not v.drawer and doesMatch(v.page, pn_or_xp) then
+            return _middle
+        elseif self:isBookmarkInPageOrder({page = pn_or_xp}, v) then
+            _end = _middle - 1
+        else
+            _start = _middle + 1
```

Comment:
This looks like it is the only binary search bit outside of ReaderAnnotation. No way to shove it into ReaderAnnotation? (I see it uses some RB related logic (isBookmarkInPageOrder), so may be not - just asking.)

Reply:
It is used for page bookmarks (dogear) only. This section of ReaderBookmark is totally devoted to dogears, so I kept them all together.